DOC-8034 -- Provide console Logging & basic troubleshooting tips

ibsoln opened the pull request:
https://issues.couchbase.com/browse/DOC-8034
https://issues.couchbase.com/browse/DOC-5786

ibsoln commented:
Hi @pasin @bmeike @Sandychuang8 @jayahariv 
I have made some changes to the logging examples in the sample code modules.
Could you please review these code changes for the appropriate languages?
Regards
Ian

bmeike commented on modules/java/examples/code_snippets/Examples.java:
```diff
@@ -748,26 +748,44 @@ public void testDatabaseEncryption() throws CouchbaseLiteException {
-        final File path = new File("/usr/local/MyApp/logs")
-        Database.log.getFile().setConfig(new LogFileConfiguration(path.toString()));
-        Database.log.getFile().setLevel(LogLevel.INFO);
+          final File path = new File("/usr/local/MyApp/logs")
```
I suggest, instead of "/usr/local", you use System.getProperty("user.dir")).  Nothing really wrong with this, as is.  On most systems, though, it probably won't actually work...

ibsoln replied:
Changed to:

`          final File path = new File(System.getProperty("user.dir").concat("/MyApp/logs"));`

Followed up in Slack https://couchbase.slack.com/archives/DS1QC61RV/p1612862225000700

bmeike commented on modules/java/pages/_partials/_attributes-module.adoc:
```diff
@@ -281,4 +281,53 @@
+
+// BEGIN Logs and logging references
+// :url-api-class-logging: {url-api-references}Logging.html[Logging classes]
+
```
Not sure what this is.... I can't actually review it...

ibsoln replied:
np

bmeike commented on modules/ROOT/pages/_partials/commons/common-logging.adoc:
```diff
@@ -29,94 +29,178 @@ include::{root-partials}_block-abstract.adoc[]
+
+[NOTE]
+--
+The retrieval of logs from the device is out of scope of this feature.
```
Boy... *That* is a shame!  Esp on Android, it is pretty hard to do.  Unless customers can get logs to us, they are essentially useless.

It is exactly the crack through which things fall: we, often "log it", as a way of addressing an inconsistency or an edge case in our design.  If we then don't provide a way of getting to the logs, well...

ibsoln replied:
What is the process for this? Is it anything like the SG_Collect https://docs.couchbase.com/sync-gateway/current/sgcollect-info.html

Will leave for now as this is partially covered in https://issues.couchbase.com/browse/DOC-6475

bmeike commented on modules/ROOT/pages/_partials/commons/common-logging.adoc:
```diff
@@ -29,94 +29,178 @@ include::{root-partials}_block-abstract.adoc[]
+--
+The retrieval of logs from the device is out of scope of this feature.
+
+ifdef::is-android[]
```
FYI as of 3.0, it will disappear.  I believe it is already deprecated...

ibsoln replied:
Thanks

bmeike commented on modules/ROOT/pages/_partials/commons/common-logging.adoc:
```diff
@@ -29,94 +29,178 @@ include::{root-partials}_block-abstract.adoc[]
+* <<lbl-console-logs>>
++
+--
+You can independently configure and control console logs, which provides a convenient method of accessing diagnostic information during debugging scenarios.
```
Would it be worth noting, here, that, on Android, the most effective strategy is to turn file logging off, altogether (ignore the warning that this produces), and to use the Console logger (which logs to logcat) exclusively.

On the Android platform, we strongly prefer the output of the console log (logcat) to the file logs.

ibsoln replied:
added at line 92 below

`TIP: It can often be more effective to just use the Console logger (which logs to logcat). 
Note, a warning is displayed when you set continuous (file) logging *off*.`

bmeike commented on modules/ROOT/pages/_partials/commons/common-logging.adoc:
```diff
@@ -29,94 +29,178 @@ include::{root-partials}_block-abstract.adoc[]
+To change default settings use database's {url-api-class-log} method to set the required values -- see <<eg-cons-log>>
+
+You will primarily use {url-api-class-log-console-getcons} and {url-api-class-log-console} to control console logging.
+
```
I don't know if this is worth noting... although the platform code will log Verbose and Debug logs, in release versions (more or less, the only thing that external users *ever* get), Core will not.  Release Core does not produce any verbose or debug logs: it will log only error, warning and info, regardless of the log-level set in the logger.

ibsoln replied:
... hm maybe TMI.  Is this across all platforms?

bmeike reviewed (approved):
LGTM.... Some random thoughts for your consideration but nothing particularly important

jayahariv commented on modules/swift/examples/code_snippets/SampleCodeTest.swift:
```diff
@@ -58,17 +58,28 @@ class SampleCodeTest {
 
+    func dontTestConsoleLogging() throws {
+        // tag::console-logging[]
+        Database.log.Console.domains = LogDomains.all // <.>
```
do we need to specify explicitly the `LogDomains.all`? also console is lowercase. 
```suggestion
        Database.log.console.domains = .all // <.>
```

jayahariv commented on modules/swift/examples/code_snippets/SampleCodeTest.swift:
```diff
@@ -58,17 +58,28 @@ class SampleCodeTest {
+    func dontTestConsoleLogging() throws {
+        // tag::console-logging[]
+        Database.log.Console.domains = LogDomains.all // <.>
+        Database.log.Console.level = LogLevel.verbose // <.>
```
?
```suggestion
        Database.log.console.level = .verbose // <.>
```

jayahariv commented on modules/swift/examples/code_snippets/SampleCodeTest.swift:
```diff
@@ -58,17 +58,28 @@ class SampleCodeTest {
     func dontTestEnableCustomLogging() throws {
         // tag::set-custom-logging[]
-        Database.log.custom = LogTestLogger()
+        Database.log.custom = LogTestLogger(LogLevel.warning) // <.>
```
? 
```suggestion
        Database.log.custom = LogTestLogger(.warning) // <.>
```

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
 
+- (void) dontTestEnableConsoleLogging {
+    // tag::console-logging[]
+    [CBLDatabase.log.console setLevel: kCBLLogDomainAll]; // <.>
```
?
```suggestion
    CBLDatabase.log.console.domains = kCBLLogDomainAll; // <.>
```

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
+- (void) dontTestEnableConsoleLogging {
+    // tag::console-logging[]
+    [CBLDatabase.log.console setLevel: kCBLLogDomainAll]; // <.>
+    [CBLDatabase.log.console setDomain: kCBLLogLevelVerbose]; // <.>
```
?
```suggestion
    CBLDatabase.log.console.level = kCBLLogLevelVerbose; // <.>
```

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
 - (void) dontTestEnableCustomLogging {
     // tag::set-custom-logging[]
     [CBLDatabase.log setCustom:[[LogTestLogger alloc] init]];
+    [CBLDatabase.log setCustom:[LogLevel CBLLogLevel.kCBLLogLevelWarning]]; // <.>
```
This is wrong! it should be either set inside the `LogTestLogger` class or pass it as a constructor parameter. Something like the above comment

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
 
 - (void) dontTestEnableCustomLogging {
     // tag::set-custom-logging[]
     [CBLDatabase.log setCustom:[[LogTestLogger alloc] init]];
```
```suggestion
    [CBLDatabase.log setCustom:[[LogTestLogger alloc] initWithLogLevel: kCBLLogLevelWarning]];
```

jayahariv reviewed (changes requested):
Could you please take care of the comments

ibsoln commented:
Thanks for the reviews @bmeike and @jayahariv. I've adopted most of the suggestions now I think, with the exception of one or two queries.

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
     // tag::set-custom-logging[]
-    [CBLDatabase.log setCustom:[[LogTestLogger alloc] init]];
+    [CBLDatabase.log setCustom:[[LogTestLogger alloc] initWithLogLevel: kCBLLogLevelWarning]];
+    [CBLDatabase.log setCustom:[LogLevel CBLLogLevel.kCBLLogLevelWarning]]; // <.>
```
?

jayahariv commented on modules/objc/examples/code_snippets/SampleCodeTest.m:
```diff
@@ -164,18 +164,31 @@ - (void) dontTestLogging {
     // tag::set-custom-logging[]
-    [CBLDatabase.log setCustom:[[LogTestLogger alloc] init]];
+    [CBLDatabase.log setCustom:[[LogTestLogger alloc] initWithLogLevel: kCBLLogLevelWarning]];
+    [CBLDatabase.log setCustom:[LogLevel CBLLogLevel.kCBLLogLevelWarning]]; // <.>
```
If we adopted the constructor to accept the logLevel, like above, this line is not necessary.
```suggestion
```

ibsoln replied:
@jayahariv Thanks. I removed that line (190) now. Can you approve so I can merge please?

Sandychuang8 commented on modules/csharp/examples/code_snippets/Program.cs:
```diff
@@ -968,27 +968,66 @@ private static void StartReplication()
         {
             // tag::set-custom-logging[]
-            Database.Log.Custom = new LogTestLogger();
+            Database.Log.Custom = new LogTestLogger(LogLevel.Warning); // <.>
```
```suggestion
            Database.Log.Custom = new LogTestLogger(); // <.>
            or
            Database.Log.Custom = new LogTestLogger { Level = LogLevel.Warning };
```
Base on the custom LogTestLogger class, there is no constructor taking any parameter, new LogTestLogger() is the correct way to init the LogTestLogger object. Or new LogTestLogger { Level = LogLevel.Warning }; to create LogTestLogger  with it's Level property set to LogLevel.Warning.

ibsoln replied:
Thank-you @Sandychuang8

ibsoln committed:
Update modules/objc/examples/code_snippets/SampleCodeTest.m

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
Update modules/objc/examples/code_snippets/SampleCodeTest.m

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
Correct type in swift code examples (line 64)

Update modules/swift/examples/code_snippets/SampleCodeTest.swift

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
Remove constructor in EnableCustomLogging

Co-authored-by: Sandychuang8 <sandy.chuang@couchbase.com>

ibsoln committed:
Update modules/swift/examples/code_snippets/SampleCodeTest.swift

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
DOC-8034 -- Provide console Logging & basic troubleshooting tips
https://issues.couchbase.com/browse/DOC-8034
https://issues.couchbase.com/browse/DOC-5786

ibsoln committed:
DOC-8034 -- Replace usr/local in Java Example

https://issues.couchbase.com/browse/DOC-8034
https://issues.couchbase.com/browse/DOC-5786

ibsoln committed:
Update modules/swift/examples/code_snippets/SampleCodeTest.swift

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
DOC-8034 -- Remove redundant code line

https://issues.couchbase.com/browse/DOC-8034

ibsoln committed:
Apply suggestions from ObjC code review

Co-authored-by: Jayahari Vavachan <10448770+jayahariv@users.noreply.github.com>

ibsoln committed:
Add tip to use just console logging (per feedback)

ibsoln committed:
Use System.getProperty("user.dir") not usr/local/
